### ISSUE TITLE
Error when trying to reconnect to a dead server

### ISSUE THREAD

**Issue by bojanbass** [luin/ioredis#141]
Hi,

today our Redis server was down (connect timeout) and this was the chance to see how reconnection logic was working in ioredis. But I noticed some strange error:

```javascript
/Projects/SRGit/nodejs/admin_web_console/node_modules/ioredis/lib/redis/event_handler.js:78
      self.connect().catch(function () {});
                          ^
TypeError: Object #<Redis> has no method 'catch'
    at null._onTimeout (/Projects/SRGit/nodejs/admin_web_console/node_modules/ioredis/lib/redis/event_handler.js:78:27)
    at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
```

Looks like calling self.connect() doesn't return a promise, but Redis instance, which is strange. Does anyone know what's wrong here?

**Comment by luin** [comment]
That's really strange. Are you able to reproduce it?

**Comment by bojanbass** [comment]
Yes, every time. Do you need any other code to check the problem?

**Comment by luin** [comment]
1. Which version of ioredis are you using?
2. Could you please add the following codes to log `self` and `self.connect` before the 78 line of event_handler.js:

  ```javascript
  console.log('print self:', self);
  console.log('print self.connect:', self.connect.toString());
  ```

**Comment by bojanbass** [comment]
Oh, you helped me to solve my problem! The problem is in how do I set up redis config. I have a file called redis.js, which has this content:

```javascript
var Redis = require('ioredis')

var redis = new Redis({
  port: 6379,          // Redis port
  host: config.redisDs,   // Redis host
  db: 5
});

module.exports = redis;

module.exports.connect = function () {
  //var redisConfig = url.parse(config.REDISURL);
  //var client = redis.createClient(redisConfig.port, redisConfig.hostname);
  //
  //if (redisConfig.auth !== null)
  //    client.auth(redisConfig.auth.split(':')[1]);
  //
  //client.on('error', function(e){
  //    console.log(e);
  //});
  return redis;
};
```

and for redis init I only include this file in my main app like this:
```javascript
require('./redis');
```

So when I added console.log(self.connect.toString()), the content returned was the one from my custom connect function, so the default one was overriden. Removing my module.exports.connect function soved the problem.

Thank you for your help!

**Comment by luin** [comment]
:laughing:

**Comment by bojanbass** [comment]
Just another minor question: in my case when redis server is down, I'm trying to catch errors when connection fails, so I added this code inside my redis.js module:

```javascript
redis.on('error', function (e) {
  console.log(e);
});
```

The strange thing is that the number of error events executions is huge, somewhere around 250 in 1 minute. Is there anything wrong with reconnection logic or is this expected behaviour?

Thanks!

**Comment by luin** [comment]
What errors did you get?

**Comment by bojanbass** [comment]
```javascript
Error: connect ECONNREFUSED
    at errnoException (net.js:905:11)
    at Object.afterConnect [as oncomplete] (net.js:896:19)
```

**Comment by luin** [comment]
That's normal because `ECONNREFUSED` is emitted when connection has failed. By default ioredis will try to reconnect immediately(see https://github.com/luin/ioredis#auto-reconnect) so you will see many errors like this.

**Comment by bojanbass** [comment]
Ok, I found out what's going on with retry logic. If I set a custom retry function like this:
```javascript
  retryStrategy: function (times) {
    return 1000;
  }
```
then retry works as expected, so 1 retry each second. But if I leave the defaults:

```javascript
retryStrategy: function (times) {
    var delay = Math.min(times * 2, 2000);
    return delay;
  }
```

then the delay calculated is a very small number, starting 2, 4, 6, 8 and so on. But this represents miliseconds, that's why it will try to reconnect multiple times per second. I think the default function is missing * 1000 multiplier for the delay variable, so reconnection intervals would then be 2, 4, 6, 8 seconds, which makes more sense to me. Or am I wrong?
